### ISSUE TITLE
Corrupted aab created sometimes

### ISSUE THREAD

**Issue by GeneralVimes** [airsdk/Adobe-Runtime-Support#1042]
@GeneralVimes weird, but good I guess ... if you do find that issue again please let me know. We should be clean-building the AAB file every time you run ADT so it seems a bit odd to be able to create a corrupted zip: if you find this again then please just send us the corrupt zip file and we'll inspect it to see if we can see a problem in the indexing or similar..

thanks

_Originally posted by @ajwfrost in https://github.com/airsdk/Adobe-Runtime-Support/discussions/994#discussioncomment-1121580_

Got it again! :)

So, in the attempt to build a minimal game's aab, I removed the Enhance ane, but kept google analytics swc. And I've got the corrupted aab. Here it is: https://www.dropbox.com/s/c4zb928fk8xsn0f/EvolutionArena-GAswc-corrupted.aab?dl=0

Also this time I've got the AndroidStudioProject folder. Where can I send it?

I'll continue experiments. For the history: today morning building an aab with both Enhance ANE and GA swc created uncorrupted aab, as well as creating aab without any ane or swc created uncorrupted aab, too

**Comment by GeneralVimes** [comment]
And now what is the strangest. I removed GA swc from the project, compiled it and built aab again. The aab is corrupted, too! And it has the same size, as the previous corrupted aab, but with GA swc.

**Comment by ajwfrost** [comment]
Hi - okay, weird but I hope we can work something out - I'll look a the AAB file...

For the AndroidStudioProject folder(s) please use:
https://transfer.harman.com/requests/eAz5vvHYibImzG6bd1ykpT

thanks

**Comment by GeneralVimes** [comment]
Next experiment: before packaging aab (without ane or swc) I deleted the previous .aab and the previous AndroidStudioProject folder. It's not corrupted! :)

**Comment by GeneralVimes** [comment]
Sent the results of 4 builds.
1. Game has Google Analytics swc. I don't remember, whether I had AndroidStudioProject, but I had previous EvolutionArena-captive-runtime.aab before buiiding the new. Resut: corrupted aab.
2. Removed GA swc from the game. Did not delete the previous AndroidStudioProject folder and EvolutionArena-captive-runtime.aab. Result: corrupted aab (of the same size as the previous one
3. Deleted the previous AndroidStudioProject folder and EvolutionArena-captive-runtime.aab. Result: not corupted aab
4. Added Google Analytics swc. Deleted the previous AndroidStudioProject folder and EvolutionArena-captive-runtime.aab. Result: not corupted aab

So, looks like keeping the previous .aab or the previous AndroidStudioProject when building .aab again, leads to the creation of corrupted aab.

**Comment by ajwfrost** [comment]
One thing to note: you're passing signing information in to the ADT command here, to create a signed AAB file (you don't actually need to, but it is/should be supported). But when I use jarsigner to verify:
```
jarsigner: java.lang.SecurityException: SHA1 digest error for BUNDLE-METADATA/com.android.tools.build.libraries/dependencies.pb
```

The ADT source code suggests:
1) if you're using the `AndroidStudioProject` folder, then this is deleted (including all contents) at the start of bundle creation. I guess if this fails then it could fail silently.. but worth checking the folder creation date/time to see whether this is the case.
2) the build process uses gradle and then as an output, picks up the generated `app-release.aab` file from the gradle build output folder
3) a) at this point, if there's no signing then we remove a previous output, and copy this file into the required filename
b) alternatively if you provided signing options, we open the .aab file and run through the contents to get all the file checksums, then write the `META-INF` signatures into the .aab. From an initial glance I can't currently see how that signed (in-memory) file gets written to the output location so we'll check this: maybe there's a problem with it at this point...
4) finally, if it was using a temporary folder (rather than `AndroidStudioProject`) then we clean that up.

thanks

**Comment by GeneralVimes** [comment]
Thank you!
I also want to add on the method with which I generate AAB. I'm working with Flash Develop, and here I described my steps: https://www.reddit.com/r/as3/comments/owag5h/building_aab_with_flash_develop_working_experience/

In the meanwhile I tried the following: I commented the line #KeepAndroidStudioOutput=true back, but I kept the old aab, and as the result the created aab became corrupted.

**Comment by GeneralVimes** [comment]
But this does not happen always. Often new aab is not corrupted. Let me try one more thing... I think it's connected with adding/removing ane or swc between compilations

**Comment by GeneralVimes** [comment]
So, here is my hypothesis now:
keeping old .aab before publishing leads to new .aab corruption if between the builds an swc or ane is removed from the project. In this case the new .aab has exactly the same size as the old .aab.

**Comment by ajwfrost** [comment]
... ah - okay so I've worked out how this is working now! If we're using the AAB signing mechanism, then we set the output file and the base classes are just opening that as a read/write archive. Then only the 'new' files from the new generated AAB are added to it via the signature generation code, so the final signature will be based upon those files. BUT if the previous output file had already got some extra contents in it (i.e. you have a library in the existing AAB that's not in the new one) then it will keep that content, but this will screw up the zip directory and the jar signature.

Wow... thanks for finding this one and for isolating the cause! Annoyingly, we only just pushed out an updated release, but we will get this fix in; for the minute though, hopefully it's okay to just manually delete the target .aab file prior to building...

thanks
